File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isLetter, isSpace, isUpper)

responseFor :: String -> String
responseFor xs
  | isQuestion s && isShouting s = "Calm down, I know what I'm doing!"
  | isQuestion s = "Sure."
  | isShouting s = "Whoa, chill out!"
  | s == "" = "Fine. Be that way!"
  | otherwise = "Whatever."
  where
    s = filter (not . isSpace) xs

isQuestion :: String -> Bool
isQuestion [] = False
isQuestion s = last s == '?'

isShouting :: String -> Bool
isShouting s = all (\c -> not (isLetter c) || isUpper c) s && any isLetter s

File: cpp/bob/bob.cpp
#include "bob.h"
#include <algorithm>

namespace bob
{

    std::string hey(const std::string &input)
    {
        bool only_whitespace = std::all_of(input.begin(), input.end(), [](unsigned char c)
                                           { return std::isspace(c); });

        if (only_whitespace)
            return "Fine. Be that way!";

        bool has_letter = false;
        bool all_letters_upper = true;

        for (unsigned char c : input)
        {
            if (std::isalpha(c))
            {
                has_letter = true;

                if (!std::isupper(c))
                {
                    all_letters_upper = false;
                }
            }
        }

        bool is_yelling = has_letter && all_letters_upper;

        auto last_non_space = std::find_if(input.rbegin(), input.rend(), [](unsigned char c)
                                           { return !std::isspace(c); });

        bool is_question = last_non_space != input.rend() && *last_non_space == '?';

        if (is_yelling && is_question)
        {
            return "Calm down, I know what I'm doing!";
        }

        if (is_question)
        {
            return "Sure.";
        }

        if (is_yelling)
        {
            return "Whoa, chill out!";
        }

        return "Whatever.";
    }

} // namespace bob
